x-pack/plugins/apm/server/routes/span_links/get_linked_children.ts

```diff
@@ -0,0 +1,133 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { rangeQuery } from '@kbn/observability-plugin/server';
+import {
+  PROCESSOR_EVENT,
+  SPAN_ID,
+  SPAN_LINKS,
+  SPAN_LINKS_TRACE_ID,
+  SPAN_LINKS_SPAN_ID,
+  TRACE_ID,
+  TRANSACTION_ID,
+} from '../../../common/elasticsearch_fieldnames';
+import { ProcessorEvent } from '../../../common/processor_event';
+import type { SpanRaw } from '../../../typings/es_schemas/raw/span_raw';
+import type { TransactionRaw } from '../../../typings/es_schemas/raw/transaction_raw';
+import { Setup } from '../../lib/helpers/setup_request';
+import { getBufferedTimerange } from './utils';
+
+async function fetchLinkedChildrenOfSpan({
+  traceId,
+  setup,
+  start,
+  end,
+  spanId,
+}: {
+  traceId: string;
+  setup: Setup;
+  start: number;
+  end: number;
+  spanId?: string;
+}) {
+  const { apmEventClient } = setup;
+
+  const { startWithBuffer, endWithBuffer } = getBufferedTimerange({
+    start,
+    end,
+  });
+
+  const response = await apmEventClient.search(
+    'fetch_linked_children_of_span',
+    {
+      apm: {
+        events: [ProcessorEvent.span, ProcessorEvent.transaction],
+      },
+      _source: [SPAN_LINKS, TRACE_ID, SPAN_ID, PROCESSOR_EVENT, TRANSACTION_ID],
+      body: {
+        size: 1000,
+        query: {
+          bool: {
+            filter: [
+              ...rangeQuery(startWithBuffer, endWithBuffer),
+              { term: { [SPAN_LINKS_TRACE_ID]: traceId } },
+              ...(spanId ? [{ term: { [SPAN_LINKS_SPAN_ID]: spanId } }] : []),
+            ],
+          },
+        },
+      },
+    }
+  );
+
+  return response.hits.hits;
+}
+
+function getSpanId(source: TransactionRaw | SpanRaw) {
+  return source.processor.event === ProcessorEvent.span
+    ? (source as SpanRaw).span.id
+    : (source as TransactionRaw).transaction?.id;
+}
+
+export async function getLinkedChildrenCountBySpanId({
+  traceId,
+  setup,
+  start,
+  end,
+}: {
+  traceId: string;
+  setup: Setup;
+  start: number;
+  end: number;
+}) {
+  const linkedChildren = await fetchLinkedChildrenOfSpan({
+    traceId,
+    setup,
+    start,
+    end,
+  });
+
+  return linkedChildren.reduce<Record<string, number>>(
+    (acc, { _source: source }) => {
+      source.span?.links?.forEach((link) => {
+        // Ignores span links that don't belong to this trace
+        if (link.trace.id === traceId) {
+          acc[link.span.id] = (acc[link.span.id] || 0) + 1;
+        }
+      });
+      return acc;
+    },
+    {}
+  );
+}
+
+export async function getLinkedChildrenOfSpan({
+  traceId,
+  spanId,
+  setup,
+  start,
+  end,
+}: {
+  traceId: string;
+  spanId: string;
+  setup: Setup;
+  start: number;
+  end: number;
+}) {
+  const linkedChildren = await fetchLinkedChildrenOfSpan({
+    traceId,
+    spanId,
+    setup,
+    start,
+    end,
+  });
+
+  return linkedChildren.map(({ _source: source }) => {
```

Comment:
I know that we've discussed this before, but AFAICT we still need to filter here. Because the request itself might return false positives, e.g. when querying for trace.id: a, span.id: a, and a span has links to trace.id:a+span.id:b and trace.id:b+span.id:a, it should not be returned here. Maybe I'm misunderstanding though?

Reply:
>Maybe I'm misunderstanding though?

I think so 😅.

The `getLinkedChildrenOfSpan` does NOT care about the `span.links` on the Transactions/Spans returned by the query. It's only interested on the Transactions/Spans `trace.id` and `transaction.id or span.id`. 

That is because when I filter by `span.links.trace.id: a` and `span.links.span.id: a` the query returns all transactions/spans that match these two values, so I just need to loop through the items returned creating children span links.

Consider the following example:
**Filter:** span.links.trace.id = D and span.links.span.id = D.
Query returns:
```js
{
  trace.id: 'A',
  span:{
    id: 'A',
    links: [
      {trace.id: 'B', span.id: 'B'},
      {trace.id: 'D', span.id: 'D'},
      {trace.id: 'D', span.id: 'D.2'},
    ]
  }
}
```

The expected return from `getLinkedChildrenOfSpan` is:
```js
[
 {trace: {id: 'A'}, span: {id: 'A'}}
]
```

Like I explained above, the function doesn't care that the returned item has other span links like (span.links.trace.id: B/span.links.span.id: B and span.links.trace.id: D/span.links.span.id: D.2). It only cares that the item returned has the span.links.trace.id and span.links.span.id I filtered.

Reply:
@cauemarcondes I understand that it doesn't care about the span links, however, this will return documents that don't actually have a span link to the parent. Documents that have span links to trace.id:A, and span links to span.id: B, but never in the same span link, are also returned.

Reply:
I'm not sure if I understood your comment.

But when calling ES with a query like this and with spanId **defined**:
```js
 filter: [
              { term: { [SPAN_LINKS_TRACE_ID]: traceId } },
              ...(spanId ? [{ term: { [SPAN_LINKS_SPAN_ID]: spanId } }] : []),
            ],
```

I think it returns only documents that both `links.trace.id` **and** `links.span.id` are equal to the values specified, am I right? It does an AND not OR. 

Maybe we should talk again, it's difficult to fully explain it here.

Reply:
ES stores arrays internally as:

- `span.id: [ 'A', 'B' ]`
- `trace.id: [ 'A', 'B' ]`

So you can't query on the combination of span.id and trace.id inside a single span link.

Reply:
Fixed it, and created a test scenario with this case https://github.com/elastic/kibana/pull/126630/commits/c189bd06173716f1d87b0a8f1c0e21b85f37727c#diff-a334ef7866737cd750655465746e6ccce271b99f9a6aa0c28aae248546479f9fR143-R144